File: app/src/main/java/de/baumann/browser/Browser/AdBlock.java
package de.baumann.browser.Browser;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import de.baumann.browser.Database.RecordAction;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.*;

public class AdBlock {
    private static final String FILE = "hosts.txt";
    private static final Set<String> hosts = new HashSet<>();
    private static final List<String> whitelist = new ArrayList<>();
    private static final Locale locale = Locale.getDefault();

    private static void loadHosts(final Context context) {
        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                AssetManager manager = context.getAssets();
                try {
                    BufferedReader reader = new BufferedReader(new InputStreamReader(manager.open(FILE)));
                    String line;
                    while ((line = reader.readLine()) != null) {
                        hosts.add(line.toLowerCase(locale));
                    }
                } catch (IOException i) {
                    Log.w("Browser", "Error loading hosts", i);
                }
            }
        });
        thread.start();
    }

    private synchronized static void loadDomains(Context context) {
        RecordAction action = new RecordAction(context);
        action.open(false);
        whitelist.clear();
        whitelist.addAll(action.listDomains());
        action.close();
    }

    private static String getDomain(String url) throws URISyntaxException {
        url = url.toLowerCase(locale);

        int index = url.indexOf('/', 8); // -> http://(7) and https://(8)
        if (index != -1) {
            url = url.substring(0, index);
        }

        URI uri = new URI(url);
        String domain = uri.getHost();
        if (domain == null) {
            return url;
        }
        return domain.startsWith("www.") ? domain.substring(4) : domain;
    }

    private final Context context;

    public AdBlock(Context context) {
        this.context = context;

        if (hosts.isEmpty()) {
            loadHosts(context);
        }
        loadDomains(context);
    }

    public boolean isWhite(String url) {
        for (String domain : whitelist) {
            if (url.contains(domain)) {
                return true;
            }
        }
        return false;
    }

    boolean isAd(String url) {
        String domain;
        try {
            domain = getDomain(url);
        } catch (URISyntaxException u) {
            return false;
        }
        return hosts.contains(domain.toLowerCase(locale));
    }

    public synchronized void addDomain(String domain) {
        RecordAction action = new RecordAction(context);
        action.open(true);
        action.addDomain(domain);
        action.close();
        whitelist.add(domain);
    }

    public synchronized void removeDomain(String domain) {
        RecordAction action = new RecordAction(context);
        action.open(true);
        action.deleteDomain(domain);
        action.close();
        whitelist.remove(domain);
    }

    public synchronized void clearDomains() {
        RecordAction action = new RecordAction(context);
        action.open(true);
        action.clearDomains();
        action.close();
        whitelist.clear();
    }
}

File: app/src/main/java/de/baumann/browser/browser/AdBlock.kt
package de.baumann.browser.browser

import kotlin.jvm.Synchronized
import de.baumann.browser.database.RecordDb
import de.baumann.browser.unit.RecordUnit
import android.annotation.SuppressLint
import android.content.Context
import de.baumann.browser.preference.ConfigManager
import de.baumann.browser.util.DebugT
import android.util.Log
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.URI
import java.net.URISyntaxException
import java.util.*
import kotlin.Throws

class AdBlock(private val context: Context): KoinComponent {
    private val config: ConfigManager by inject()

    fun isWhite(url: String?): Boolean {
        for (domain in whitelist) {
            if (url != null && url.contains(domain!!)) {
                return true
            }
        }
        return false
    }

    fun isAd(url: String): Boolean {
        val domain: String = try {
            getDomain(url).lowercase(locale)
        } catch (u: URISyntaxException) {
            return false
        }
        return hosts.contains(domain) || isAdExtraSites(url)
    }

    private fun isAdExtraSites(url: String): Boolean {
        return config.adSites.any { url.contains(it, true) }
    }

    @Synchronized
    fun addDomain(domain: String?) {
        with(RecordDb(context)) {
            open(true)
            addDomain(domain, RecordUnit.TABLE_WHITELIST)
            close()
        }
        whitelist.add(domain)
    }

    @Synchronized
    fun removeDomain(domain: String?) {
        with(RecordDb(context)) {
            open(true)
            deleteDomain(domain, RecordUnit.TABLE_WHITELIST)
            close()
        }
        whitelist.remove(domain)
    }

    @Synchronized
    fun clearDomains() {
        with(RecordDb(context)) {
            open(true)
            clearDomains()
            close()
        }
        whitelist.clear()
    }

    private fun loadHosts(context: Context) {
        val thread = Thread {
            val debugT = DebugT("loadHosts")
            val manager = context.assets
            try {
                val reader = BufferedReader(InputStreamReader(manager.open(FILE)))
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    line?.let { hosts.add(it.lowercase(locale)) }
                }
            } catch (i: IOException) {
                Log.w("browser", "Error loading hosts", i)
            }
            debugT.printTime()
        }
        thread.start()
    }


    companion object {
        private const val FILE = "hosts.txt"
        private val hosts: MutableSet<String> = HashSet()
        private val whitelist: MutableList<String?> = ArrayList()

        @SuppressLint("ConstantLocale")
        private val locale = Locale.getDefault()
        @Synchronized
        private fun loadDomains(context: Context) {
            val action = RecordDb(context)
            action.open(false)
            whitelist.clear()
            whitelist.addAll(action.listDomains(RecordUnit.TABLE_WHITELIST))
            action.close()
        }

        @Throws(URISyntaxException::class)
        private fun getDomain(url: String): String {
            var url = url
            url = url.lowercase(locale)
            val index = url.indexOf('/', 8) // -> http://(7) and https://(8)
            if (index != -1) {
                url = url.substring(0, index)
            }
            val uri = URI(url)
            val domain = uri.host ?: return url
            return if (domain.startsWith("www.")) domain.substring(4) else domain
        }
    }

    init {
        if (hosts.isEmpty()) {
            loadHosts(context)
        }
        loadDomains(context)
    }
}